### PR TITLE
Fix handling MBEDTLS_ERR_SSL_WANT_READ with mbedtls functions

### PR DESCRIPTION
**Pull request by mazimkhan**
mbedtls functions that may return MBEDTLS_ERR_SSL_WANT_WRITE or MBEDTLS_ERR_SSL_WANT_READ should be called again until there is an error or they succeed.

### CODE REVIEW

**Comment by mazimkhan** [comment]
@yanesca Please review.

File: tls-client/main.cpp [hunk 1]
```diff
@@ -132,6 +133,8 @@ class HelloHTTPS {
         mbedtls_x509_crt_free(&_cacert);
         mbedtls_ssl_free(&_ssl);
         mbedtls_ssl_config_free(&_ssl_conf);
+        _tcpsocket->close();
```
Comment (mazimkhan) [inline tls-client/main.cpp]: ```close()``` is a harmless method and if you see its implementation, it checks if the socket is open or not. Moving it to the destructor add a lot of simplicity as we don't have to call it every time we ```return``` from ```startTest``` or use ```goto``` statements.

File: tls-client/main.cpp [hunk 2]
```diff
@@ -216,23 +219,21 @@ class HelloHTTPS {
...
+            ret = mbedtls_ssl_handshake(&_ssl);
+        } while (ret != 0 && (ret == MBEDTLS_ERR_SSL_WANT_READ ||
+                ret == MBEDTLS_ERR_SSL_WANT_WRITE));
         if (ret < 0) {
-            if (ret != MBEDTLS_ERR_SSL_WANT_READ &&
-                ret != MBEDTLS_ERR_SSL_WANT_WRITE) {
-                print_mbedtls_error("mbedtls_ssl_handshake", ret);
-                onError(_tcpsocket, -1 );
-            }
+            print_mbedtls_error("mbedtls_ssl_handshake", ret);
```
Comment (mazimkhan) [inline tls-client/main.cpp]: Since ```close``` is moved to the destructor we don't need call to ```onError```. Major reason of not calling ```onError``` is that it misleads in the logs that error has occurred due to Socket.

Comment (mazimkhan) [inline tls-client/main.cpp]: It applies to all comments for same reason below.

File: tls-client/main.cpp [hunk 3]
```diff
@@ -301,7 +301,7 @@ class HelloHTTPS {
     /**
      * Closes the TCP socket
      */
-    void close() {
+    void close() { // Do we need this functions?
```
Comment (mazimkhan) [inline tls-client/main.cpp]: Agreed and removed.

File: tls-client/main.cpp (hunk 1 above)
Comment (yanesca) [inline tls-client/main.cpp]: My main reason for suggesting this change wasn't some effective gain on the robustness of the application, I merely prefer that solution because of the logical coherence of the code.

File: tls-client/main.cpp (hunk 2 above)
Comment (yanesca) [inline tls-client/main.cpp]: If we leave closing the socket to the destructor, then we have a different problem: the `onError()` function becomes obsolete and we need to remove it and replace each call with the line `_error = true;`.

Comment (mazimkhan) [inline tls-client/main.cpp]: member ```_error```, functions ```error()``` and ```done``` aren't used anywhere. Hence setting it also doesn't make sense. looks more like incomplete code. I would rather remove it.

File: tls-client/main.cpp (hunk 1 above)
Comment (mazimkhan) [inline tls-client/main.cpp]: I see more lines of code to manage closing the socket in ```startTest()``` than putting it in the constructor. We have to see trade off between coherence and pollution in code. For coherence  we should also return from a single point in ```startTest```.

Comment (yanesca) [inline tls-client/main.cpp]: Ok, then let's do that: move the code that opens the connection to the constructor!

File: tls-client/main.cpp (hunk 2 above)
Comment (yanesca) [inline tls-client/main.cpp]: Great! Then let's remove them!

**Comment by mazimkhan** [comment]
@yanesca Incorporated your comments and also made some improvements for mbedtls_ssl_read/write calls.

File: tls-client/main.cpp [hunk 4]
```diff
@@ -256,22 +261,31 @@ class HelloHTTPS {
...
+        int read = 0;
         do {
-            ret = mbedtls_ssl_read(&_ssl, (unsigned char *) _buffer, sizeof(_buffer));
-        } while (ret != 0 && (ret == MBEDTLS_ERR_SSL_WANT_READ ||
-                ret == MBEDTLS_ERR_SSL_WANT_WRITE));
+            ret = mbedtls_ssl_read(&_ssl, (unsigned char *) _buffer + read, 
+                                   sizeof(_buffer) - 1);
+            if (ret > 0)
+              read += ret;
+            else if (ret != MBEDTLS_ERR_SSL_WANT_READ &&
```
Comment (yanesca) [inline tls-client/main.cpp]: Why did you extract this condition from the head of the loop?

File: tls-client/main.cpp [hunk 5]
```diff
@@ -256,22 +261,31 @@ class HelloHTTPS {
...
         }
-        _bpos = static_cast<size_t>(ret);
+        _bpos = static_cast<size_t>(read);
 
         _buffer[_bpos] = 0;
-
-        /* Check each of the flags */
-        _got200 = _got200 || strstr(_buffer, HTTPS_OK_STR) != NULL;
-        _gothello = _gothello || strstr(_buffer, HTTPS_HELLO_STR) != NULL;
+
```
Comment (yanesca) [inline tls-client/main.cpp]: There are whitespace characters on this empty line, could you please remove them?

File: tls-client/main.cpp [hunk 6]
```diff
@@ -401,12 +412,11 @@ class HelloHTTPS {
     void onError(TCPSocket *s, int error) {
         printf("MBED: Socket Error: %d\r\n", error);
         s->close();
-        _error = true;
     }
 
 protected:
     TCPSocket* _tcpsocket;
-
+
```
Comment (yanesca) [inline tls-client/main.cpp]: There are whitespace characters on this empty line, could you please remove them?

File: tls-client/main.cpp [hunk 7]
```diff
@@ -148,9 +151,7 @@ class HelloHTTPS {
         _error = false;
         _disconnected = false;
         _request_sent = false;
-        /* Fill the request buffer */
-        _bpos = snprintf(_buffer, sizeof(_buffer) - 1, "GET %s HTTP/1.1\nHost: %s\n\n", path, HTTPS_SERVER_NAME);
-
+
```
Comment (yanesca) [inline tls-client/main.cpp]: There are whitespace characters on this empty line, could you please remove them?

File: tls-client/main.cpp [hunk 8]
```diff
@@ -216,23 +213,31 @@ class HelloHTTPS {
...
-            }
+            print_mbedtls_error("mbedtls_ssl_handshake", ret);
+            _tcpsocket->close();
             return;
         }
 
-        ret = mbedtls_ssl_write(&_ssl, (const unsigned char *) _buffer, _bpos);
+        /* Fill the request buffer */
+        _bpos = snprintf(_buffer, sizeof(_buffer) - 1, "GET %s HTTP/1.1\nHost: %s\n\n", path, HTTPS_SERVER_NAME);
+
```
Comment (yanesca) [inline tls-client/main.cpp]: There are whitespace characters on this empty line, could you please remove them?

File: tls-client/main.cpp [hunk 9]
```diff
@@ -401,12 +412,11 @@ class HelloHTTPS {
     void onError(TCPSocket *s, int error) {
```
Comment (yanesca) [inline tls-client/main.cpp]: This function has only a single call on line 210 and the string it prints conveys no additional information. I think the call on line 210 can be replaced with `_tcpsocket->close();` and this function can be deleted.

File: tls-client/main.cpp [hunk 10]
```diff
@@ -224,15 +219,25 @@ class HelloHTTPS {
...
+            _tcpsocket->close();
             return;
         }
 
+        /* Fill the request buffer */
+        _bpos = snprintf(_buffer, sizeof(_buffer) - 1, "GET %s HTTP/1.1\nHost: %s\n\n", path, HTTPS_SERVER_NAME);
+        
+        int written = 0;
         do {
             ret = mbedtls_ssl_write(&_ssl, (const unsigned char *) _buffer, _bpos);
```
Comment (yanesca) [inline tls-client/main.cpp]: The [manual for this function](https://tls.mbed.org/api/ssl_8h.html) says that it has to be called with updated arguments in case the write was partial.

File: tls-client/main.cpp [hunk 11]
```diff
@@ -256,22 +261,31 @@ class HelloHTTPS {
...
+                                   sizeof(_buffer) - 1);
+            if (ret > 0)
+              read += ret;
+            else if (ret != MBEDTLS_ERR_SSL_WANT_READ &&
+                ret != MBEDTLS_ERR_SSL_WANT_WRITE)
+              break;
+            /* Check each of the flags */
+            _got200 = _got200 || strstr(_buffer, HTTPS_OK_STR) != NULL;
+            _gothello = _gothello || strstr(_buffer, HTTPS_HELLO_STR) != NULL;
+        } while (!_got200 || !_gothello);
```
Comment (mazimkhan) [inline tls-client/main.cpp]: Because we don't know how much to read before checking for these strings or how long to run the read loop.

File: tls-client/main.cpp (hunk 4 above)
Comment (mazimkhan) [inline tls-client/main.cpp]: Because this condition is not sufficient as the loop condition. ```ret``` can have positive values. But I can squeeze the new condition in while if you want.

File: tls-client/main.cpp [hunk 12]
```diff
@@ -224,15 +219,25 @@ class HelloHTTPS {
...
+        _bpos = snprintf(_buffer, sizeof(_buffer) - 1, "GET %s HTTP/1.1\nHost: %s\n\n", path, HTTPS_SERVER_NAME);
+        
+        int written = 0;
         do {
             ret = mbedtls_ssl_write(&_ssl, (const unsigned char *) _buffer, _bpos);
-        } while (ret != 0 && (ret == MBEDTLS_ERR_SSL_WANT_READ ||
-                ret == MBEDTLS_ERR_SSL_WANT_WRITE));
+            if (ret > 0)
+              written += ret;
+            else if (ret != MBEDTLS_ERR_SSL_WANT_READ &&
```
Comment (mazimkhan) [inline tls-client/main.cpp]: Because this condition is not sufficient as the loop condition. ```ret``` can have positive values. But I can squeeze the new condition in while if you want.

File: tls-client/main.cpp (hunk 10 above)
Comment (mazimkhan) [inline tls-client/main.cpp]: Updated

File: tls-client/main.cpp [hunk 13]
```diff
@@ -256,22 +263,32 @@ class HelloHTTPS {
...
+            ret = mbedtls_ssl_read(&_ssl, (unsigned char *) _buffer + offset, 
+                                   sizeof(_buffer) - offset - 1);
+            if (ret > 0)
+              offset += ret;
+
+            /* Check each of the flags */
+            _buffer[offset] = 0;
+            _got200 = _got200 || strstr(_buffer, HTTPS_OK_STR) != NULL;
+            _gothello = _gothello || strstr(_buffer, HTTPS_HELLO_STR) != NULL;
+        } while (offset < sizeof(_buffer) && (!_got200 || !_gothello) &&
```
Comment (yanesca) [inline tls-client/main.cpp]: Here we can omit the `offset < sizeof(_buffer)` condition, because `mbedtls_ssl_read` should check it and throw an error if the data does not fit.

File: tls-client/main.cpp [hunk 14]
```diff
@@ -256,22 +263,32 @@ class HelloHTTPS {
...
+                (ret || ret == MBEDTLS_ERR_SSL_WANT_READ ||
```
Comment (yanesca) [inline tls-client/main.cpp]: I know that the functionality is the same, but would you mind if we would write (ret > 0 instead of (ret here for better readability?

File: tls-client/main.cpp (hunk 2 above)
Comment (yanesca) [inline tls-client/main.cpp]: It is removed in the latest version.

File: tls-client/main.cpp (hunk 1 above)
Comment (yanesca) [inline tls-client/main.cpp]: In the latest version the constructor remained the same and the function closes the connection before exiting and this approach resolves my above concern as well.

**Comment by yanesca** [comment]
Looks good to me!

**Comment by mazimkhan** [comment]
@andresag01 @sbutcher-arm Please review and merge.

File: tls-client/main.cpp [hunk 15]
```diff
@@ -256,22 +263,32 @@ class HelloHTTPS {
...
+            if (ret > 0)
+              offset += ret;
+
+            /* Check each of the flags */
+            _buffer[offset] = 0;
+            _got200 = _got200 || strstr(_buffer, HTTPS_OK_STR) != NULL;
+            _gothello = _gothello || strstr(_buffer, HTTPS_HELLO_STR) != NULL;
+        } while ( (!_got200 || !_gothello) &&
+                (ret > 0 || ret == MBEDTLS_ERR_SSL_WANT_READ ||
+                ret == MBEDTLS_ERR_SSL_WANT_WRITE));
```
Comment (andresag01) [inline tls-client/main.cpp]: Why are these loops here instead of the `ssl_send()` and `ssl_recv()` functions? It seems that we are having loops over and over. Perhaps it is better to just put them in one place?

File: tls-client/main.cpp [hunk 16]
```diff
@@ -210,29 +207,39 @@ class HelloHTTPS {
         ret = _tcpsocket->connect(_domain, _port);
         if (ret != NSAPI_ERROR_OK) {
             mbedtls_printf("Failed to connect\r\n");
-            onError(_tcpsocket, -1);
+            printf("MBED: Socket Error: %d\r\n", ret);
+            _tcpsocket->close();
```
Comment (mazimkhan) [inline tls-client/main.cpp]: For the first comment. Having a status print in ```main()``` deviates from the issue this PR addresses. Secondly that would make this example look more like a test case. Success or failure is apparent by looking the progress in the logs.

Second comment: putting ```_tcpsocket->close()``` in destructor was a good idea. But @yanesca wanted to maintain coherence by also moving ```_tcpsocket->connect()``` in the constructor. That I don't agree with that as connection failure will result in an object that isn't in the expected state.

Comment (mazimkhan) [inline tls-client/main.cpp]: Hence instead I called it at each exit point. This applies to all similar comments.

File: tls-client/main.cpp (hunk 15 above)
Comment (mazimkhan) [inline tls-client/main.cpp]: Since mbedtls lib expects and handles non blocking IO. The ssl_send and recv routines should not be blocking in terms of looping until data is received. Secondly in some cases like call to ```mbedtls_ssl_read```  the application does not know how much it want to receive and it sends buffer size instead of expected message size. Hence putting such loops in those routines can cause app to hang.

File: tls-client/main.cpp [hunk 17]
```diff
@@ -375,6 +387,7 @@ class HelloHTTPS {
         if(NSAPI_ERROR_WOULD_BLOCK == recv){
             return MBEDTLS_ERR_SSL_WANT_READ;
         }else if(recv < 0){
+            mbedtls_printf("Socket recv error %d\r\n", recv);
             return -1;
```
Comment (mazimkhan) [inline tls-client/main.cpp]: Did not intend to change this. This could be pat of a different PR after analysing why it was done like this originally.

File: tls-client/main.cpp [hunk 18]
```diff
@@ -390,20 +403,15 @@ class HelloHTTPS {
         size = socket->send(buf, len);
 
         if(NSAPI_ERROR_WOULD_BLOCK == size){
-            return len;
+            return MBEDTLS_ERR_SSL_WANT_WRITE;
         }else if(size < 0){
+            mbedtls_printf("Socket send error %d\r\n", size);
             return -1;
```
Comment (mazimkhan) [inline tls-client/main.cpp]: same as above.

**Comment by mazimkhan** [comment]
@andresag01 cleaned up unused ```_error``` and unused methods. Please reply to the comments.

**Comment by andresag01** [comment]
I have manually merged these changes into development branch. Refer to https://github.com/ARMmbed/mbed-os-example-tls/commit/4459c46e976ccf754f75e0028d8eb2a4041a2463

**Comment by mazimkhan** [comment]
@andresag01 @sbutcher-arm why is this not merged into master?

**Comment by andresag01** [comment]
@mazimkhan we are planning to do some additional work on this example. It was decided that we will continue in a development branch and then integrate into master once we have decided the example is stable.

**Comment by mazimkhan** [comment]
Ok but don't understand this "once we have decided the example is stable."

My understanding was that changes planned in the development branch were for minor refactoring of this example. The example has been made stable in this PR i.e. 53. The refactoring shouldn't be related or prompted by this PR i.e. could have been done on the code before.
